de_web_plugin.cpp

```diff
@@ -9781,6 +9781,50 @@ Sensor *DeRestPluginPrivate::getSensorNodeForAddressAndEndpoint(const deCONZ::Ad
 
 }
 
+/*! Compare addresses where either NWK or MAC address might not be known.
+    \returns true if both adresses have same MAC address (strong guaranty).
+    \returns true if at least one of the addresses doesn't have MAC but the NWK addresses are equal.
+    \returns false otherwise.
+*/
+bool isSameAddress(const deCONZ::Address &a, const deCONZ::Address &b)
+{
+    if (a.hasExt() && b.hasExt())
+    {
+        // nested if statement, so the NWK check won't be made if both MAC addresses are known
+        if (a.ext() != b.ext()) 
+        {
+             return false;
+        }
+    }
+    else  if (a.hasNwk() && b.hasNwk())
+    {
+       if (a.nwk() != b.nwk())
+       {
+            return false;
+       }
+    }
+    else { return false; }
+
+    return true;
+}
+
+/*! Returns the first Sensor for its given \p Address and \p Endpoint and \p Cluster or nullptr if not found.
+ */
+Sensor *DeRestPluginPrivate::getSensorNodeForAddressEndpointAndCluster(const deCONZ::Address &addr, quint8 ep, quint16 cluster)
+{
+    for (Sensor &sensor: sensors)
+    {
+        if (sensor.deletedState() != Sensor::StateNormal || !sensor.node())                             { continue; }
+        if (!isSameAddress(sensor.address(), addr))                                                     { continue; }
+        if (sensor.fingerPrint().hasInCluster(cluster) || sensor.fingerPrint().hasOutCluster(cluster))  { }
+        else                                                                                            { continue; }
+
+        if (sensor.fingerPrint().endpoint != ep)    { continue; }
+        else                                        { return &sensor; }
```

Comment:
Cool I can understand it now much quicker :+1: 

The function is called quite often, to tickle a bit more performance out, I'd suggest the following reordering:

```
if (sensor.deletedState() != Sensor::StateNormal || !sensor.node())                             { continue; }
if (sensor.fingerPrint().endpoint != ep)                                                        { continue; }
if (!isSameAddress(sensor.address(), addr))                                                     { continue; }
if (sensor.fingerPrint().hasInCluster(cluster) || sensor.fingerPrint().hasOutCluster(cluster))  { }
else                                                                                            { continue; }

return &sensor; 
```

Here the checks are ordered by "cost", the first two checks are likely in the same cache line and very fast, the remaining ones reach out to memory access since address and clusters are located somewhere else on the heap :)

Reply:
Well, I guess you know better than I do which order might be faster 😉 However, I reacall there was a reason behind it putting at least the EP check as last one. I've to admit that might have lost relevance as the initial idea was also to be able to cover switches exposing only one resource and the code changed based on our various discussions around it.

For that reason, I'd like to put some thoughts around this again to be sure.

Reply:
I think the position of the EP check is independent, if the check fails the sensor will always be skipped :)

Reply:
Yeah. But regardless, there's mostly a reason why I put stuff in there 😆 

Found the very first version, there it made some sense... Anyway, as this has all changed, the order indeed doesn't matter.